File: lib/src/test/kotlin/com/hdmain/tcpduplex/TcpDuplexTest.kt
package com.hdmain.tcpduplex

import com.hdmain.tcpduplex.protocol.Protocol
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.net.ServerSocket
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

class TcpDuplexTest {
    @Test
    fun connSendReceive() {
        val ss = ServerSocket(0)
        ss.use { serverSocket ->
            val srvResult = ArrayBlockingQueue<Result<Conn>>(1)

            thread {
                val raw = serverSocket.accept()
                try {
                    srvResult.put(Result.success(TcpDuplex.serveConn(raw)))
                } catch (e: Exception) {
                    srvResult.put(Result.failure(e))
                }
            }

            val cli = TcpDuplex.dial("127.0.0.1:${serverSocket.localPort}")
            try {
                val res = srvResult.poll(10, TimeUnit.SECONDS)!!
                val srvConn = res.getOrThrow()
                try {
                    val msg = "hello duplex".toByteArray()
                    cli.send(msg)
                    val got = srvConn.receive()
                    assertArrayEquals(msg, got)

                    val reply = "ack".toByteArray()
                    srvConn.send(reply)
                    val got2 = cli.receive()
                    assertArrayEquals(reply, got2)

                    srvConn.close()
                    assertThrows<Exception> {
                        cli.receive()
                    }
                } finally {
                    runCatching { srvConn.close() }
                }
            } finally {
                cli.close()
            }
        }
    }

    @Test
    fun connConcurrentSendReceive() {
        val ss = ServerSocket(0)
        ss.use { serverSocket ->
            val errRef = java.util.concurrent.atomic.AtomicReference<Throwable?>(null)
            val done = CountDownLatch(1)

            thread {
                try {
                    val raw = serverSocket.accept()
                    TcpDuplex.serveConn(raw).use { srv ->
                        repeat(50) { i ->
                            val b = srv.receive()
                            srv.send(b)
                        }
                    }
                } catch (e: Exception) {
                    errRef.set(e)
                } finally {
                    done.countDown()
                }
            }

            TcpDuplex.dial("127.0.0.1:${serverSocket.localPort}").use { cli ->
                repeat(50) { i ->
                    val payload = byteArrayOf(i.toByte())
                    cli.send(payload)
                    val out = cli.receive()
                    assertEquals(1, out.size)
                    assertEquals(i.toByte(), out[0])
                }
            }

            done.await(30, TimeUnit.SECONDS)
            errRef.get()?.let { throw it }
        }
    }

    @Test
    fun dialUnsupportedProtocolVersion() {
        val cfg = Config.defaultConfig()
        cfg.protocolVersion = 42u
        assertThrows<Protocol.UnsupportedVersionException> {
            TcpDuplex.dial("127.0.0.1:1", cfg)
        }
    }

    @Test
    fun pskHandshake() {
        val ss = ServerSocket(0)
        ss.use { serverSocket ->
            val cfg = Config.defaultConfig()
            cfg.handshake.preSharedKey = "unit-test-psk".toByteArray()

            thread {
                val raw = serverSocket.accept()
                TcpDuplex.serveConn(raw, cfg).use { srv ->
                    val msg = srv.receive()
                    assertArrayEquals("ping".toByteArray(), msg)
                    srv.send("pong".toByteArray())
                }
            }

            TcpDuplex.dial("127.0.0.1:${serverSocket.localPort}", cfg).use { cli ->
                cli.send("ping".toByteArray())
                val out = cli.receive()
                assertArrayEquals("pong".toByteArray(), out)
            }
        }
    }

    @Test
    fun receiveAfterPeerClose() {
        val ss = ServerSocket(0)
        val done = CountDownLatch(1)
        thread {
            try {
                val raw = ss.accept()
                TcpDuplex.serveConn(raw).use { it.close() }
            } finally {
                done.countDown()
            }
        }

        TcpDuplex.dial("127.0.0.1:${ss.localPort}").use { cli ->
            done.await(10, TimeUnit.SECONDS)
            assertThrows<Exception> { cli.receive() }
        }
        ss.close()
    }
}

File: tcpduplex/tcpduplex_test.go
package tcpduplex

import (
	"context"
	"errors"
	"net"
	"sync"
	"testing"
)

func TestConnSendReceive(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()

	type srvResult struct {
		conn *Conn
		err  error
	}
	srvCh := make(chan srvResult, 1)

	go func() {
		c, err := ln.Accept()
		if err != nil {
			srvCh <- srvResult{err: err}
			return
		}
		sc, err := ServeConn(c)
		srvCh <- srvResult{conn: sc, err: err}
	}()

	cli, err := Dial(ln.Addr().String())
	if err != nil {
		t.Fatal(err)
	}
	defer cli.Close()

	res := <-srvCh
	if res.err != nil {
		t.Fatalf("server: %v", res.err)
	}
	srvConn := res.conn
	defer func() { _ = srvConn.Close() }()

	msg := []byte("hello duplex")
	if err := cli.Send(msg); err != nil {
		t.Fatal(err)
	}
	got, err := srvConn.Receive()
	if err != nil {
		t.Fatal(err)
	}
	if string(got) != string(msg) {
		t.Fatalf("payload mismatch: %q", got)
	}

	reply := []byte("ack")
	if err := srvConn.Send(reply); err != nil {
		t.Fatal(err)
	}
	got2, err := cli.Receive()
	if err != nil {
		t.Fatal(err)
	}
	if string(got2) != string(reply) {
		t.Fatalf("reply mismatch: %q", got2)
	}

	if err := srvConn.Close(); err != nil {
		t.Fatal(err)
	}

	if _, err := cli.Receive(); err == nil {
		t.Fatal("expected error after peer close")
	}
}

func TestConnConcurrentSendReceive(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()

	errCh := make(chan error, 1)
	var wg sync.WaitGroup
	wg.Add(1)

	go func() {
		defer wg.Done()
		c, err := ln.Accept()
		if err != nil {
			errCh <- err
			return
		}
		srv, err := ServeConn(c)
		if err != nil {
			errCh <- err
			return
		}
		defer srv.Close()
		for i := 0; i < 50; i++ {
			b, err := srv.Receive()
			if err != nil {
				errCh <- err
				return
			}
			if err := srv.Send(b); err != nil {
				errCh <- err
				return
			}
		}
	}()

	cli, err := Dial(ln.Addr().String())
	if err != nil {
		t.Fatal(err)
	}
	defer cli.Close()

	for i := 0; i < 50; i++ {
		payload := []byte{byte(i)}
		if err := cli.Send(payload); err != nil {
			t.Fatal(err)
		}
		out, err := cli.Receive()
		if err != nil {
			t.Fatal(err)
		}
		if len(out) != 1 || out[0] != byte(i) {
			t.Fatalf("round-trip %d: got %v", i, out)
		}
	}

	wg.Wait()
	select {
	case err := <-errCh:
		if err != nil {
			t.Fatal(err)
		}
	default:
	}
}

func TestDialUnsupportedProtocolVersion(t *testing.T) {
	cfg := DefaultConfig()
	cfg.ProtocolVersion = 42
	_, err := DialContext(context.Background(), "127.0.0.1:1", cfg)
	if err == nil {
		t.Fatal("expected error")
	}
	if !errors.Is(err, ErrUnsupportedProtocol) {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestPSKHandshake(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()

	cfg := DefaultConfig()
	cfg.Handshake.PreSharedKey = []byte("unit-test-psk")

	go func() {
		c, err := ln.Accept()
		if err != nil {
			return
		}
		srv, err := ServeConnContext(context.Background(), c, cfg)
		if err != nil {
			t.Error(err)
			return
		}
		defer srv.Close()
		msg, err := srv.Receive()
		if err != nil || string(msg) != "ping" {
			t.Errorf("srv recv %v %q", err, msg)
			return
		}
		_ = srv.Send([]byte("pong"))
	}()

	cli, err := DialContext(context.Background(), ln.Addr().String(), cfg)
	if err != nil {
		t.Fatal(err)
	}
	defer cli.Close()

	if err := cli.Send([]byte("ping")); err != nil {
		t.Fatal(err)
	}
	out, err := cli.Receive()
	if err != nil || string(out) != "pong" {
		t.Fatalf("cli recv %v %q", err, out)
	}
}

func TestSendContextCanceled(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		c, err := ln.Accept()
		if err != nil {
			return
		}
		srv, err := ServeConn(c)
		if err != nil {
			return
		}
		defer srv.Close()
		if _, err := srv.Receive(); err != nil {
			return
		}
	}()

	cli, err := Dial(ln.Addr().String())
	if err != nil {
		t.Fatal(err)
	}

	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	if err := cli.SendContext(ctx, []byte("x")); !errors.Is(err, context.Canceled) {
		t.Fatalf("expected canceled, got %v", err)
	}

	_ = cli.Send([]byte("flush"))
	_ = cli.Close()
	wg.Wait()
}

func TestReceiveAfterPeerClose(t *testing.T) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}

	done := make(chan struct{})
	go func() {
		defer close(done)
		c, err := ln.Accept()
		if err != nil {
			return
		}
		sc, err := ServeConn(c)
		if err != nil {
			return
		}
		_ = sc.Close()
	}()

	cli, err := Dial(ln.Addr().String())
	if err != nil {
		t.Fatal(err)
	}
	defer cli.Close()

	<-done
	_ = ln.Close()

	if _, err := cli.Receive(); err == nil {
		t.Fatal("expected error after peer closed")
	}
}
